cmd/crank/beta/validate/manager.go

```diff
@@ -257,15 +280,169 @@ func (m *Manager) cacheDependencies() error {
+		return v.GetDependencies(), nil
+	case *metav1.Provider:
+		return v.GetDependencies(), nil
+	case *metav1.Function:
+		return v.GetDependencies(), nil
+	default:
+		return nil, errors.New("unknown package type")
+	}
+}
+
+func findPackageYamlType(meta []byte) (interface{}, error) {
+	// Define a list of possible types
+	candidates := []interface{}{
+		&metav1.Configuration{},
+		&metav1.Provider{},
+		&metav1.Function{},
+	}
+
+	// Try to unmarshal into each type
+	for _, candidate := range candidates {
+		if err := yaml.Unmarshal(meta, candidate); err == nil {
+			// If successful, return the candidate
+			return candidate, nil
+		}
+	}
+
+	return nil, errors.New("cannot unmarshal dependency YAML")
+}
+
+// getAndExtractPackageWithType gets the package image from cache or remote
+// and returns the package with its type (provider/function/configuration).
+func (m *Manager) getAndExtractPackageWithType(image string) (interface{}, error) {
+	// If package is already cached, get it from cache
+	path, _, meta, err := m.getPackageFromCache(image)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting package from cache for %s", image)
+	}
+
+	// If package is not cached, download and cache it
+	if path != "" {
+		_, meta, err = m.downloadAndCachePackage(image)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error downloading and caching package for %s", image)
+		}
+	}
+
+	// Find type of the package (provider/function/configuration) and return
+	return findPackageYamlType(meta)
+}
+
+// getAndExtractAllPackagesWithType gets the package images from cache or remote,
+// assigns them to the manager's deps map and returns the map with their types.
+func (m *Manager) getAndExtractAllPackagesWithType() (interface{}, error) {
+	pkgs := make(map[string]interface{})
+
+	// Getting images that are not assigned
+	for img, data := range m.deps {
+		if data == nil {
+			pkgs[img] = nil
+		}
+	}
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(pkgs))
+
+	// Get and extract packages concurrently
+	for image := range pkgs {
+		wg.Add(1)
+		go func(image string) {
+			defer wg.Done()
+
+			m.mu.Lock()
+			pkg := m.deps[image]
+			m.mu.Unlock()
+			if pkg == nil {
+				var err error
+				pkg, err = m.getAndExtractPackageWithType(image)
+				if err != nil {
+					errChan <- err
+					return
+				}
+
+				m.mu.Lock()
+				m.deps[image] = pkg
+				m.mu.Unlock()
+			}
+		}(image)
+	}
+
+	wg.Wait()
+	close(errChan)
+
+	for err := range errChan {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pkgs, nil
+}
+
+func (m *Manager) getPackageFromCache(image string) (string, [][]byte, []byte, error) {
+	path, err := m.cache.Exists(image) // returns the path if the image is not cached
+	if err != nil {
+		return path, nil, nil, errors.Wrapf(err, "cannot check if cache exists for %s", image)
+	}
+	if path == "" {
+		switch cache := m.cache.(type) {
+		case *LocalCache:
+			pkg := filepath.Join(cache.CacheDir(image), "package.yaml")
```

Comment:
Actually this was the only thing I "had" to do in this PR because I couldnt add a new method to [Cache](https://github.com/crossplane/crossplane/blob/ebe648363be45bbd9693e2d31a37b9a0486a4ce0/cmd/crank/beta/validate/cache.go#L31) interface since it already had 5 methods and adding a new one gave `interface bloat` error when I ran earthly. As a workaround I did it like this but now that I thought more on this, I think changing Exists interface method to a method named Get and handling the existance check etc. inside this method would be a good way to tackle this without compromising the abstraction. I will work on this and update this PR with a new logic.